### PR TITLE
(DO NOT MERGE) Branch for Megan to experiment with F1 thresholding.

### PR DESCRIPTION
**Pull request by samuelklee**

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ScoreVariantAnnotations.java [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+        final File variantTypeAnnotationsFile = LabeledVariantAnnotationsData.subsetAnnotationsToTemporaryFile(annotationNames, allAnnotations, isVariantType);
+        final File variantTypeScoresFile = IOUtils.createTempFile("temp", ".scores.hdf5");
+        variantTypeScorer.score(variantTypeAnnotationsFile, variantTypeScoresFile);
+        final double[] variantTypeScores = VariantAnnotationsScorer.readScores(variantTypeScoresFile);
+        final Iterator<Double> variantTypeScoresIterator = Arrays.stream(variantTypeScores).iterator();
+        IntStream.range(0, allScores.size()).filter(isVariantType::get).forEach(i -> allScores.set(i, variantTypeScoresIterator.next()));
+    }
+
+    @Override
+    void writeExtractedVariantToVCF(final VariantContext vc,
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ScoreVariantAnnotations.java:260]: @meganshand you'll want to modify this VCF-writing method to perform hard filtering as desired. Currently, we don't actually do any hard filtering; we just take the appropriate score(s) for each variant from a global iterator, use a truth-sensitivity converter (which is built from the appropriate set of truth scores) to convert the score (or the maximum of scores, in AS mode) to a sensitivity, and then write both to the VCF.

Seems like you could add whatever tool-level parameters you'll need to specify the cutoff calculation, and then store the resulting threshold globally. And perhaps do this separately for SNP/INDEL.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ScoreVariantAnnotations.java [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+                    : null;
+        }
+
+        // TODO validate modes
+        if (snpScorer == null && indelScorer == null) {
+            throw new UserException.BadInput("At least one serialized scorer must be present in model files.");
+        }
+
+        // TODO extract method and constants
+        final File snpTruthScores = new File(modelPrefix + ".snp.truthScores.hdf5");
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/ScoreVariantAnnotations.java:152]: @meganshand here's where you can read in the truth scores for each variant type and do whatever calculations you need. Check the VariantAnnotationsScorer interface for the `static double[] readScores(final File inputFile)` method, which you can use to read in the scores.

**Comment by samuelklee** [comment]
Hmm, as we discussed, I'm not sure this will be so straightforward, given that we only have easy access to scores for positive truth---and hence, no false positives, which precludes calculation of precision and F1. I *think* we could pass a VCF for a sample with gold-standard positives and negatives and use the existing code for extracting labels, but this will require a bit of engineering and be more trouble than it's worth.

There are other options---see https://ir.cwi.nl/pub/30479, for example. We might want to experiment with the LL score discussed there before moving on to their methods for estimating F1. Doing a literature search for other discussions of optimizing F1 or other metrics in the context of positive-unlabeled learning might be worthwhile, but I think most methods will probably involve some sort of estimation of the base rate in unlabeled data.

Perhaps @davidbenjamin has some thoughts? I see some related stuff going on in ThresholdCalculator, but I have to admit that I can't tell whether that's used in a similar PU context. Also note that depending on the model used, we might not have well calibrated posteriors---the IsolationForest simply outputs scores in a unit interval, for example.

**Comment by samuelklee** [comment]
Just fixed a minor bug. Shouldn't affect scores or anything of import---just corrected some behavior in the scoring step when passing labels for extracted variants in positive/negative mode. So maybe worth pulling in the fix whenever you start playing around with that.

**Comment by samuelklee** [comment]
Hmm, I started taking a stab at the LL score implementation, but I think it's going to complicate the code quite a bit and add some branching options to the tool interfaces. Compounding this with a change in the use of "truth" and "validation" terminology, I fear that the resulting differences from the legacy strategy might be a bit much for users to digest!

So I'd want to better understand the cost/benefit before we proceed. How critical is automatic tuning of the hard threshold? And what's the relative importance to method changes that increase AUC (i.e., as opposed to figuring out where on the curve to hard threshold)? Is there a clear path forward for evaluating such a tuning process? @meganshand would be glad to chat more!

**Comment by samuelklee** [comment]
One more thing: I'm also wondering if it would be possible to get a quick, preliminary evaluation of such a process without actually doing the work of adding it into the training tool. It's probably possible to do a slightly more "manual" validation split (say, using one or a few chromosomes), run the score tool on that validation set, use some external code to calculate the desired threshold from the resulting scores, and then use that threshold going forward.

Actually, now that I've written it out, that sounds a lot cleaner and more flexible! Let me try to hack together the corresponding workflow.

**Comment by samuelklee** [comment]
Running something like this on the outputs of the score tool from a validation shard could help determine the threshold. Below I show the results of running with a test VCF of 50 1000G WES samples. Note that "training" and "truth" are used in the VQSR meanings, and that we derive positive labels for the validation set from their union; but one could imagine variations on this.

```
import h5py
import numpy as np
import matplotlib.pyplot as plt

annot_file = 'test.all-unlabeled.annot.hdf5'
scores_file = 'test.all-unlabeled.scores.hdf5'

with h5py.File(annot_file, 'r') as f:
    is_training_n = f['/labels/training'][()].astype(bool)
    is_truth_n = f['/labels/truth'][()].astype(bool)
    
with h5py.File(scores_file, 'r') as f:
    score_n = f['/data/scores'][()]

score_sort_order_n = score_n.argsort()
sorted_score_n = score_n[score_sort_order_n]
is_positive_n = is_training_n | is_truth_n

p_n = is_positive_n[score_sort_order_n]
m_n = ~p_n

tp_n = np.cumsum(p_n[::-1])[::-1]
fn_n = np.cumsum(p_n)
mp_n = np.cumsum(m_n[::-1])[::-1]

recall_n = tp_n / (tp_n + fn_n)
LL_score_n = recall_n**2 / (mp_n / sum(m_n))

LL_score_argmax = sorted_score_n[np.argmax(LL_score_n)]
recall_at_LL_score_argmax = recall_n[np.argmax(LL_score_n)]

plt.plot(sorted_score_n, LL_score_n, label=f'LL score, argmax = {LL_score_max:.2f}')
plt.plot(sorted_score_n, recall_n, label=f'recall at LL score argmax = {recall_at_LL_score_argmax:.2f}')
plt.xlabel('score')
plt.legend()
plt.show()
```

image

Would be interesting to also plot e.g. F1 vs. score, where F1 is determined on a orthogonal set of positive/negative gold-standard labels (e.g., GIAB), to see how close the LL score determined on the "training/truth" labels gets us.

**Comment by meganshand** [comment]
Happy to chat more @samuelklee, but for now in our use case this isn't super important to have this tuning step fully integrated into the tool. Of course I didn't respond to your first comment quickly enough and I completely agree with the approach you laid out in the second comment. 

I can try adding this new workflow to the WDL and testing it out as well. I have some GIAB samples so the test you mention about comparing the LL to an orthogonal F1 should be doable.

**Comment by samuelklee** [comment]
@meganshand Here's a quick example:

image
image

Note that the malaria data used was pretty small: chr1-2 training (~20k positive training/truth variants, ~50k negative training variants---although note the negative threshold was not tuned), chr3 validation (~50k variants), and chr4-6 test (~150k variants). The LL score is calculated from using the training/truth positives used in the filtering tool, while the F1 score is calculated using "orthogonal truth" positives/negatives determined using 3 families of ~30 trios each. There's some arbitrariness in the boundary used to delineate the latter; I'd expect using gold-standard GIAB truth would be even better.

Not sure how much we can conclude, but that the validation and test F1 are similar and the validation LL score isn't too far off is encouraging. I would hope that with more variants (as in your data), things might stabilize even more. That said, we should expect some discrepancy between LL and F1, according to one of the papers linked above.

Happy to chat more about how you might implement this in your WDL---should be pretty straightforward!

**Comment by samuelklee** [comment]
Great! And yes, LL will be optimized separately for SNPs and INDELs.

How about this for a first workflow to target?

1) Run ExtractVariantAnnotations on a training set of chromosomes. You can keep training/truth labels as in Best Practices, for now.
2) Run TrainVariantAnnotationsModel on that. We'll use the truth scores generated here for any sensitivity conversions---i.e., we'll be calibrating scores only to the truth sites that are contained in the training set of chromosomes.
3) Use the trained model to run a single shard of ScoreVariantAnnotations on a validation set of chromosomes.
4) Run some variation of the above script on the resulting outputs to determine SNP and INDEL score thresholds for optimizing the corresponding LL scores. We can also add some code to the script to use the truth scores from step 2 to convert these score thresholds into truth-sensitivity thresholds.
5) Provide these truth-sensitivity thresholds to ScoreVariantAnnotations and use them to hard filter. Evaluate on a test set of chromosomes.

If all looks good, we can later move steps 3-4 into the train tool and automate the passing of sensitivities in 5 via outputs in the model directory. This will let us keep the basic interface of ScoreVariantAnnotations the same, but we'll have to add a few basic parameters to TrainVariantAnnotationsModel to control the train/validation split.

So I think all this branch is missing is step 5---we'll simply need to add command-line parameters for the SNP/INDEL sensitivity thresholds and then do the hard filtering in the VCF writing method highlighted above. Do you think you can handle implementing that in this branch, and then the rest at the WDL level? I can help with the python script for the LL stuff (or anything else), if needed.

Not sure if you got a chance to check out what your collaborators are doing in the methods you're looking to compare against, but it would be good to understand if this basic scheme for train/validation/test splitting can be replicated over there. You'll want to compare apples to apples, after all!

**Comment by samuelklee** [comment]
Some tweaking of those aforementioned rules (to make them much more conservative about using the trios to infer condition positive---namely, by requiring >30% of each column be green, i.e., non-reference Mendelian consistent) brings the F1 and LL into much better agreement:

image
image

image
image

**Comment by samuelklee** [comment]
Think it might be worth saving a VariantFiltration pass for the bit of code it'd take, but up to you!

ScoreVariantAnnotations will output both the raw "VQSLOD" score and the converted sensitivity, so we're free to specify thresholds on either. However, given that different types of models may have scores in different ranges (e.g., BGMM vs. IsolationForest, positive/negative vs. positive-only, etc.), I think it's better to restrict  all command-line options to be expressed in terms of a sensitivity.

Same thing goes if you decide to filter externally with VariantFiltration for now. Even though you have both quantities available to you, just use the sensitivity.

This brings us to questions related to whether we want to keep the old VQSR requirements of having both training and truth sets specified. For example, we could instead drop the distinction between training and truth for the new tools, and always calibrate sensitivity to the training set (you can essentially force this behavior with the current code by specifying training=true,truth=true for all of your resources).

**Comment by meganshand** [comment]
Sorry to get back to this thread so much later, but I'm finally working on this now. Here are some initial plots of the LL score on chr7 (after being trained on chr1-6) using the branch in this PR:

image
image

I haven't compared this to the F1 score yet, but that's my next step.

**Comment by meganshand** [comment]
Looking at the LL score cutoff from the plot above (around -.35) shows that the recall is expected to be quite low at this cutoff. This is also the case when looking at NA12878 with GiaB truth as is seen in the ROC plot below. The group of red points to the right is the region where the F1 score is maximized, while the purple points to the left represent the cutoff using the LL score. It looks like the LL score is much too conservative and chooses a sensitivity around .4 even when the overall ROC curve looks pretty good and a more optimal tradeoff exists. 

image

For now we can afford to use a held out truth sample to tune this threshold so I don't think we need to use the LL score at this point in time.

It would still be good to look into using the positive and negative model (this version only used the positive model) and plot the F1 and LL scores on the same plot to dig into what's going on.

**Comment by samuelklee** [comment]
Might've been a typo in the script for calculating LL above, mea culpa! I think we should actually have:

```
p_n = is_positive_n[score_sort_order_n]
M = len(p_n)
M1_n = np.arange(M)[::-1] + 1

tp_n = np.cumsum(p_n[::-1])[::-1]
fn_n = np.cumsum(p_n)

recall_n = tp_n / (tp_n + fn_n)
LL_score_n = recall_n**2 / (M1_n / M)
```

Waiting for @meganshand to check behavior on her data (and she should probably check my math, as well!), but here's what I see now for the malaria SNPs:

image

That we're now coming in on the other side of F1 at higher recall is encouraging, but this may still be too conservative. In any case, there's hopefully less variance in this quantity w.r.t. F1 across our datasets than is caused by typos... :man_facepalming: :
